File: DynamicProgramming/2D_DP/GridDP/Triangle.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int nr = triangle.size();
        vector<vector<int>> dp;
        for (int i = 0; i < nr; i++) {
            dp.push_back(vector<int>(i+1, 0));
        }
        dp[0][0] = triangle[0][0];
        for (int r = 1; r < nr; r++) {
            for (int c = 0 ; c < r+1; c++) {
                if (c==0) {
                    dp[r][c] = dp[r-1][c] + triangle[r][c];
                }
                else if (c==r) {
                    dp[r][c] = dp[r-1][c-1] + triangle[r][c];
                }
                else {
                    dp[r][c] = min(dp[r-1][c], dp[r-1][c-1]) + triangle[r][c];
                }
            }
        }
        auto ans = min_element(dp[nr-1].begin(), dp[nr-1].end());
        return *ans;
    }
};

File: DynamicProgramming/2D_DP/Grid_PathDP/Triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        nr = len(triangle)
        dp = [[0 for n in range(i + 1)] for i in range(nr)]
        dp[0][0] = triangle[0][0]
        for r in range(1, nr):
            for c in range(r+1):
                if c == 0:
                    dp[r][c] = dp[r-1][c] + triangle[r][c]
                elif c == r:
                    dp[r][c] = dp[r-1][c-1] + triangle[r][c]
                else:
                    dp[r][c] = min(dp[r-1][c], dp[r-1][c-1]) + triangle[r][c]
        return min(dp[nr-1])

